woocommerce/woocommerce-gutenberg-products-block
# Expose cart errors as notices

We identified an issue that API responses that returned errors were being stored to the data stores, but not converted to notices.

To resolve this, I've converted `receiveError` to a thunk that creates error notices, following the pattern of `receiveCart`/`notifyQuantityChanges` that @opr implemented.

When detangling this problem I converted a few files to Typescript and merged some hooks. Ill add notes in the diff.

Fixes #7721

#### Other Checks


### Testing

#### Automated Tests

#### User Facing Testing

We need to test this by throwing errors from the API. Add some error code here: https://github.com/woocommerce/woocommerce-blocks/blob/95e460409816f5dcd896cd81c5c7ab100ef5f817/src/StoreApi/Routes/V1/CartSelectShippingRate.php#L74

You can try:

- `die( 1 );` to test a generic error
- `throw new RouteException( 'test-error', 'Test', 400 );` to test a named error

1. With the above in place, on both cart and checkout, select a shipping rate
2. You should see the error notice appear at the top of the cart/checkout

There is no other error handling or retry mechanism beyond this (out of scope).

Since this revises the changes in https://github.com/woocommerce/woocommerce-blocks/pull/7363, repeat these testing steps also:

1. Add two items to your cart, item A and item B.
2. Go to the Cart block, and ensure you see no errors.
3. Open a new tab, and edit product A. Set it to be out of stock.
4. Without reloading the page return to the Cart block, and increase the quantity of any product. Ensure an error appears.
5. Edit product A again, and set it to be in stock.
6. Without reloading the page return to the Cart block, and increase the quantity of any product. Ensure the previous error is removed.
7. In separate tabs, edit product A and B. Set them both to be out of stock.
8. Without reloading the page return to the Cart block, and increase the quantity of any product. Ensure an error appears showing both products are out of stock.
9. Edit only one product, set it back to in stock.
10. Without reloading the page return to the Cart block, and increase the quantity of any product. Ensure the error stays, but that it only shows that one of the products is out of stock.


### WooCommerce Visibility
